sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/SynchronousMessageSubscriber.java

```diff
@@ -118,128 +144,192 @@ private void drainQueue() {
+
+                        return difference < 0 ? 0 : difference;
+                    });
+
+
+                    currentWork = workQueue.poll();
+                    continue;
                 }
-            });
+
+                final SynchronousReceiveWork work = currentWork;
+                logger.verbose("workId[{}] numberOfEvents[{}] Current work updated.", work.getId(),
+                    work.getNumberOfEvents());
+
+                work.start();
+                requestUpstream(work.getNumberOfEvents());
+
+                return work;
+            }
+
+            return currentWork;
+        }
     }
 
     /**
-     * {@inheritDoc}
+     * Adds additional credits upstream if {@code numberOfMessages} is greater than the number of {@code REQUESTED}
+     * items.
+     *
+     * @param numberOfMessages Number of messages required downstream.
      */
-    @Override
-    protected void hookOnError(Throwable throwable) {
-        logger.error("[{}] Errors occurred upstream", currentWork.getId(), throwable);
-        synchronized (currentWorkLock) {
-            currentWork.error(throwable);
+    private void requestUpstream(long numberOfMessages) {
+        if (isTerminated()) {
+            logger.info("Cannot request more messages upstream. Subscriber is terminated.");
+            return;
+        }
+
+        final Subscription subscription = UPSTREAM.get(this);
+        if (subscription == null) {
+            logger.info("There is no upstream to request messages from.");
+            return;
         }
-        dispose();
+
+        final long currentRequested = REQUESTED.get(this);
+        final long difference = numberOfMessages - currentRequested;
+
+        logger.verbose("Requesting messages from upstream. currentRequested[{}] numberOfMessages[{}] difference[{}]",
+            currentRequested, numberOfMessages, difference);
+
+        if (difference <= 0) {
+            return;
+        }
+
+        Operators.addCap(REQUESTED, this, difference);
+
+        subscription.request(difference);
     }
 
     @Override
-    protected void hookOnCancel() {
+    public void dispose() {
+        super.dispose();
+
+        dispose("Upstream completed the receive work.", null);
+    }
+
+    private void dispose(String message, Throwable throwable) {
+        super.dispose();
+
         if (isDisposed.getAndSet(true)) {
             return;
         }
 
         synchronized (currentWorkLock) {
             if (currentWork != null) {
-                currentWork.complete();
+                currentWork.complete(message, throwable);
+                currentWork = null;
             }
-            if (currentTimeoutOperation != null && !currentTimeoutOperation.isDisposed()) {
-                currentTimeoutOperation.dispose();
+
+            SynchronousReceiveWork w = workQueue.poll();
+            while (w != null) {
+                w.complete(message, throwable);
+                w = workQueue.poll();
             }
```

Comment:
nit; maybe we don't have put the code that signal completion to `workQueue` works inside the lock, but the current apporach won't hurt given the object is being disposed of, and this is a one-time route. I was just marking the observation (in case we revisit this code later, look up history)